### PR TITLE
Makefile: compile decode|encode|vpp|transcode if --enablecapi

### PR DESCRIPTION
**Pull request by wudping**
Compile yami[decode|encode|vpp|transcode] together with --enablecapi

To fix the bug: 
Enabling capi and tests should not skip yami[decode|encode|vpp|transcode] compilation 

Signed-off-by: wudping <dongpingx.wu@intel.com>

### CODE REVIEW

**Comment by xuguangxin** [comment]
-1, our c api and c++ api have different memory layout, final address this problem need remove the constructor from decoder/encoder.

**Comment by wudping** [comment]
@xuguangxin , okay, I will update.

**Comment by xuguangxin** [comment]
Hi @wudping 
Any update on this?

**Comment by wudping** [comment]
Hi @xuguangxin ,
I am working on this now.  The yamidecode can use --enable-capi option, but yamitranscode is still need more effort. I will push the patch in this week.

**Comment by xuguangxin** [comment]
Hi @wudping 
We are not suppose have --enable-capi in compile time any more. We should have --capi option in yamidecode and yamitranscode in run time.

**Comment by wudping** [comment]
okay @xuguangxin , I get it.

File: tests/decode.cpp [hunk 1]
```diff
@@ -33,25 +30,35 @@
...
+            fprintf(stderr, "createInput failed, cannot convert VppInput to VppInputDecode\n");
+            return input;
+        }
+        if (!inputDecode->config(*display)) {
+            input.reset();
+            fprintf(stderr, "VppInputDecode config failed.\n");
+            return input;
+        }
+    }else{
+        SharedPtr<VppInputDecode> inputDecode = std::tr1::dynamic_pointer_cast<VppInputDecode>(input);
```
Comment (xuguangxin) [inline tests/decode.cpp]: Totally duplicate code with if branch, Do not repeat yourself is important  please refer to https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

File: tests/decodehelp.cpp [hunk 2]
```diff
@@ -48,20 +49,28 @@ static void printHelp(const char* app)
     printf("      3: texture: export video frame as drm name (RGBX) + texture from drm name\n");
     printf("      4: texture: export video frame as dma_buf(RGBX) + texutre from dma_buf\n");
     printf("      5: texture: export video frame as dma_buf(NV12) + texture from dma_buf. not implement yet\n");
+    printf("  --use-capi: use the codec capi to encode or decode, default(false)\n");
```
Comment (xuguangxin) [inline tests/decodehelp.cpp]: just name it as --capi, and group it after -m.

File: tests/vppoutputencode.cpp [hunk 3]
```diff
@@ -126,6 +126,11 @@ static void setEncodeParam(const SharedPtr<IVideoEncoder>& encoder,
     encoder->setParameters(VideoConfigTypeAVCStreamFormat, &streamFormat);
 }
 
+VppOutputEncode::VppOutputEncode()
+{
+    m_outputBuffer.format = OUTPUT_BUFFER_LAST;
```
Comment (xuguangxin) [inline tests/vppoutputencode.cpp]: why we need change this?

File: tests/decodehelp.cpp (hunk 2 above)
Comment (wudping) [inline tests/decodehelp.cpp]: okay, thanks @xuguangxin .

File: tests/decode.cpp (hunk 1 above)
Comment (wudping) [inline tests/decode.cpp]: thanks @xuguangxin , but the member function config() is not a virtual function.

File: tests/vppoutputencode.cpp (hunk 3 above)
Comment (wudping) [inline tests/vppoutputencode.cpp]: m_outputBuffer.format has a initial value before this patch when --disable-capi
`#ifndef __ENABLE_CAPI__		
      VideoEncOutputBuffer():data(0), bufferSize(0), dataSize(0)		
     , remainingSize(0), flag(0),  **format(OUTPUT_BUFFER_LAST)**, timeStamp(0) {		
     };		
 #endif`

It's okay to delete this piece of code. I had tested this patch without this.

File: tests/decode.cpp (hunk 1 above)
Comment (xuguangxin) [inline tests/decode.cpp]: We can make code better like this:
```
if(para.useCAPI) {
    SharedPtr<VppInputDecodeCapi> inputDecode = std::tr1::dynamic_pointer_cast<VppInputDecodeCapi>(input);		       
    if (inputDecode && inputDecode->config(*display))
          return input;
} else {
....
}
input.reset();
fprintf(stderr, "invalid input file.\n");
```

File: tests/decodehelp.cpp [hunk 4]
```diff
@@ -92,6 +101,11 @@ bool processCmdLine(int argc, char** argv, DecodeParameter* parameters)
         case 'o':
             outputFile = optarg;
             break;
+        case 0:
+            switch (option_index) {
+            case 1:
+                parameters->useCAPI = strncasecmp(optarg, "true", 4) == 0;
+                break;
```
Comment (xuguangxin) [inline tests/decodehelp.cpp]: wired, is it build able

File: tests/decode.cpp (hunk 1 above)
Comment (wudping) [inline tests/decode.cpp]: okay, thanks @xuguangxin .

File: tests/decodehelp.cpp (hunk 4 above)
Comment (wudping) [inline tests/decodehelp.cpp]: yes, it is. I have included the head file "#include <getopt.h>".

Comment (xuguangxin) [inline tests/decodehelp.cpp]: ok, i see you removed the { at line 73, could you not remove it?

File: tests/decodehelp.cpp [hunk 5]
```diff
@@ -92,6 +101,11 @@ bool processCmdLine(int argc, char** argv, DecodeParameter* parameters)
         case 'o':
             outputFile = optarg;
             break;
+        case 0:
+            switch (option_index) {
```
Comment (xuguangxin) [inline tests/decodehelp.cpp]: this { will match with L112, and they are not in same indent level.

Comment (wudping) [inline tests/decodehelp.cpp]: oh, you are right, I am sorry. Thanks @xuguangxin, I will update it.

File: egl/egl_util.h [hunk 6]
```diff
@@ -19,7 +19,7 @@
 #include <EGL/egl.h>
 #include <EGL/eglext.h>
 
-#include "VideoCommonDefs.h"
+#include <VideoCommonDefs.h>
```
Comment (xuguangxin) [inline egl/egl_util.h]: Please use Libyami.h and LibyamiC.h for all c++ and c place.

**Comment by wudping** [comment]
covered by regression test

**Pull request merged**